### PR TITLE
Refactor purging logic and overhaul the residuals computation.

### PR DESCRIPTION
**Pull request by vdbergh**
The computation of bad_workers is now done inside purge_run(), where it logically belongs.
    
As a side effect of the refactoring it is now easy to have different purging criteria for auto purging and manual purging.
    
As a proof of concept we have set the purging criteria for manual purging to p=0.01 and residual=4.5, whereas for auto purging (disabled by default) we have set them to p=0.001 and residual=7.0.
    
We also use a different logic for computing residuals because the previous residuals (defined as the maximum of the absolute values of the Pearson adjusted residuals in a row) have an unknown distribution. Now we compute (conceptually) for every row in the chi2 table the chi2 value of the 2x3 table obtained by that row and the sum of the other rows. This chi2 value follows a chi2 distribution with two degrees of freedom (under the null hypothesis that the workers are homogeneous). To keep some compatibility with the previous residuals, we convert the chi2 value to a standard normal value.
    
To give an indication of the severity of a residual, Fishtest uses a coloring scheme. However this coloring scheme was created when far fewer workers were active. These days most of the yellow and red residuals are just the result of statistical fluctuations and as a result nobody pays any attention to them.
    
In this PR we update the coloring scheme. More precisely we color residuals according to the p=0.05 and p=0.01 thresholds, but using the Bonferroni correction, which divides these p-values by the number of workers.

### CODE REVIEW

**Comment by ppigazzini** [comment]
DEV updated.

**Comment by vdbergh** [comment]
Sigh: this one gives an exception

https://dfts-0.pigazzini.it/tests/view/6142f13a171fa5e08b4ec4f3

**Comment by vdbergh** [comment]
Well it seems to have been correctly purged, but it showed an exception.

**Comment by ppigazzini** [comment]
```
Sep 26 08:33:58 vps1572612706 pserve[32432]:     response = view(request)
Sep 26 08:33:58 vps1572612706 pserve[32432]:   File "/home/usr00/fishtest/server/fishtest/views.py", line 907, in tests_purge
Sep 26 08:33:58 vps1572612706 pserve[32432]:     purged = request.rundb.purge_run(run, p=0.01, res=4.5)
Sep 26 08:33:58 vps1572612706 pserve[32432]:   File "/home/usr00/fishtest/server/fishtest/rundb.py", line 1066, in purge_run
Sep 26 08:33:58 vps1572612706 pserve[32432]:     fishtest.stats.stat_util.update_SPRT(results, run["args"]["sprt"])
Sep 26 08:33:58 vps1572612706 pserve[32432]: NameError: name 'results' is not defined
```

**Comment by vdbergh** [comment]
Ok I see. I did not try and SPRT runs. I'll fix it.

**Comment by ppigazzini** [comment]
I got another exception simply opening this test, perhaps related
https://dfts-0.pigazzini.it/tests/view/614ad9ac7bdc23e77ceb88c3

```
Sep 26 08:43:45 vps1572612706 pserve[32432]:   File "/home/usr00/fishtest/server/env/lib/python3.8/site-packages/mako/runtime.py", line 920, in _render_context
Sep 26 08:43:45 vps1572612706 pserve[32432]:     _exec_template(inherit, lclcontext, args=args, kwargs=kwargs)
Sep 26 08:43:45 vps1572612706 pserve[32432]:   File "/home/usr00/fishtest/server/env/lib/python3.8/site-packages/mako/runtime.py", line 947, in _exec_template
Sep 26 08:43:45 vps1572612706 pserve[32432]:     callable_(context, *args, **kwargs)
Sep 26 08:43:45 vps1572612706 pserve[32432]:   File "/home/usr00/fishtest/server/fishtest/templates/base.mak", line 125, in render_body
Sep 26 08:43:45 vps1572612706 pserve[32432]:     <main>${self.body()}</main>
Sep 26 08:43:45 vps1572612706 pserve[32432]:   File "/home/usr00/fishtest/server/fishtest/templates/tests_view.mak", line 322, in render_body
Sep 26 08:43:45 vps1572612706 pserve[32432]:     <td style="background-color:${task['residual_color']}">${f"{task['residual']:.3f}"}</td>
Sep 26 08:43:45 vps1572612706 pserve[32432]: TypeError: unsupported format string passed to dict.__format__
```

**Comment by vdbergh** [comment]
The second one is troublesome. There is complaint about missing keys.

The error message seems to say that `task['residual']` is a dictionary, but I can't see how that would be possible.

**Comment by ppigazzini** [comment]
Master crashed too with that test (and a few others).
I reloaded a fresh DB, DEV updated, now that test is OK.
I will be away from the keyboard for 90 minutes.

**Comment by ppigazzini** [comment]
Bad data in the official DB is my nightmare :)

**Comment by vdbergh** [comment]
> Bad data in the official DB is my nightmare :)

Actually the second issue makes sense. With the "sane residuals" PR `task["residuals"]` was indeed a dictionary. In master (and the current PR) it is a float.

I guess this is a lesson not to make db format changes which are incompatible with master....

The current PR makes no db format changes.

**Comment by ppigazzini** [comment]
DEV updated

**Comment by vdbergh** [comment]
A histogram of residuals of a random `100000x3` contigency table. Pretty normal!

normal

**Comment by ppigazzini** [comment]
DEV updated.

**Comment by vdbergh** [comment]
Thanks! Next step: using the Bonferroni correction for colouring (possible now that we know the distribution of the residuals).

**Comment by ppigazzini** [comment]
DEV updated , loaded the latest DB backup.

Some couples of tests:

https://tests.stockfishchess.org/tests/view/614da58d7bdc23e77ceb8a7e
https://dfts-0.pigazzini.it/tests/view/614da58d7bdc23e77ceb8a7e

https://tests.stockfishchess.org/tests/view/6150e7b70d0a3569cbc64167
https://dfts-0.pigazzini.it/tests/view/6150e7b70d0a3569cbc64167

https://tests.stockfishchess.org/tests/view/6151d0510d0a3569cbc641f3
https://dfts-0.pigazzini.it/tests/view/6151d0510d0a3569cbc641f3

**Comment by ppigazzini** [comment]
@vdbergh I'm scheduling some time odds tests with a fleet of normal workers and one bad worker with the scaling factor divided by N (hacked by hand). With N>2 after some games the residuals of the bad worker suddenly are disappeared, see
https://dfts-0.pigazzini.it/tests/view/6153048423f9171afc4010e9
https://dfts-0.pigazzini.it/tests/view/6152fc3e37c5c049af8a867b

**Comment by vdbergh** [comment]
Hmmm. All workers in these tests have enough games. So they should all have residuals. Apparently one of the workers is filtered.

The chi2 test is recomputed each time the page is opened. I made a version that prints out the content of the dictionary `users` before and after filtering and the matrix `observed`, after filtering. It would be nice to know if the filtering was done correctly.

**Comment by ppigazzini** [comment]
- https://dfts-0.pigazzini.it/tests/view/6153048423f9171afc4010e9
```
aggregated users= {'7a73b420-35ac-4493-a00a-a5735c5fbfb6': [2282.0, 2174.0, 5868.0], 'aaad230f-7349-4119-b8fb-aa3db9ec4dfa': [233.0, 181.0, 354.0], 'abc7e11e-fb5f-4674-9f26-9ac11eec9e09': [161.0, 130.0, 217.0], 'efec9f1b-fec5-4c08-bcd4-11522d67744f': [125.0, 127.0, 236.0], 'edecf068-ecf5-4f14-85ff-969c82462f68': [227.0, 242.0, 435.0], '8f205586-b189-4929-bf52-c5f7d6754f8c': [127.0, 139.0, 270.0], '6bc81708-e173-435d-926f-d3030cbac13b': [136.0, 134.0, 250.0], '68c087da-91bb-46ca-bc22-15aee89605c9': [236.0, 234.0, 414.0], '2cc84465-c146-490e-b427-588fd6c500e6': [164.0, 148.0, 296.0], 'ddfa7fa9-f5fa-4af8-a528-d9b731748920': [136.0, 130.0, 250.0], '44eb5a5d-3acc-40b5-84f9-39feae3c9926': [234.0, 232.0, 434.0], '23c0da7c-1e3f-4031-8b96-abc864b253c3': [131.0, 132.0, 241.0], 'e4aa6bc4-cc33-4778-ac4a-93da227ba08c': [238.0, 213.0, 397.0], '0c21d656-b287-4e52-a50f-22818402b919': [154.0, 154.0, 284.0], 'b07f1601-9e0f-4d80-8098-12d57e10b51c': [137.0, 128.0, 215.0], '47437c91-b2a0-4e6d-b65f-998a6af3383b': [238.0, 220.0, 354.0], '64572e51-4a6d-4cbc-b02d-ee077a2948e3': [235.0, 208.0, 377.0]}
users after filtering= {'7a73b420-35ac-4493-a00a-a5735c5fbfb6': [2282.0, 2174.0, 5868.0], 'aaad230f-7349-4119-b8fb-aa3db9ec4dfa': [233.0, 181.0, 354.0], 'abc7e11e-fb5f-4674-9f26-9ac11eec9e09': [161.0, 130.0, 217.0], 'efec9f1b-fec5-4c08-bcd4-11522d67744f': [125.0, 127.0, 236.0], 'edecf068-ecf5-4f14-85ff-969c82462f68': [227.0, 242.0, 435.0], '8f205586-b189-4929-bf52-c5f7d6754f8c': [127.0, 139.0, 270.0], '6bc81708-e173-435d-926f-d3030cbac13b': [136.0, 134.0, 250.0], '68c087da-91bb-46ca-bc22-15aee89605c9': [236.0, 234.0, 414.0], '2cc84465-c146-490e-b427-588fd6c500e6': [164.0, 148.0, 296.0], 'ddfa7fa9-f5fa-4af8-a528-d9b731748920': [136.0, 130.0, 250.0], '44eb5a5d-3acc-40b5-84f9-39feae3c9926': [234.0, 232.0, 434.0], '23c0da7c-1e3f-4031-8b96-abc864b253c3': [131.0, 132.0, 241.0], 'e4aa6bc4-cc33-4778-ac4a-93da227ba08c': [238.0, 213.0, 397.0], '0c21d656-b287-4e52-a50f-22818402b919': [154.0, 154.0, 284.0], 'b07f1601-9e0f-4d80-8098-12d57e10b51c': [137.0, 128.0, 215.0], '47437c91-b2a0-4e6d-b65f-998a6af3383b': [238.0, 220.0, 354.0], '64572e51-4a6d-4cbc-b02d-ee077a2948e3': [235.0, 208.0, 377.0]}
observed=
 [[2282. 2174. 5868.]
 [ 233.  181.  354.]
 [ 161.  130.  217.]
 [ 125.  127.  236.]
 [ 227.  242.  435.]
 [ 127.  139.  270.]
 [ 136.  134.  250.]
 [ 236.  234.  414.]
 [ 164.  148.  296.]
 [ 136.  130.  250.]
 [ 234.  232.  434.]
 [ 131.  132.  241.]
 [ 238.  213.  397.]
 [ 154.  154.  284.]
 [ 137.  128.  215.]
 [ 238.  220.  354.]
 [ 235.  208.  377.]]
```
- https://dfts-0.pigazzini.it/tests/view/6152fc3e37c5c049af8a867b
```
aggregated users= {'efec9f1b-fec5-4c08-bcd4-11522d67744f': [158.0, 139.0, 243.0], 'aaad230f-7349-4119-b8fb-aa3db9ec4dfa': [248.0, 226.0, 426.0], 'b07f1601-9e0f-4d80-8098-12d57e10b51c': [197.0, 150.0, 313.0], '38684b64-4680-4982-b22e-43d8343beaff': [1676.0, 1713.0, 4795.0], '0c21d656-b287-4e52-a50f-22818402b919': [187.0, 184.0, 349.0], '8f205586-b189-4929-bf52-c5f7d6754f8c': [164.0, 169.0, 327.0], 'ddfa7fa9-f5fa-4af8-a528-d9b731748920': [204.0, 159.0, 297.0], 'abc7e11e-fb5f-4674-9f26-9ac11eec9e09': [165.0, 144.0, 291.0], '23c0da7c-1e3f-4031-8b96-abc864b253c3': [178.0, 168.0, 314.0], '6bc81708-e173-435d-926f-d3030cbac13b': [170.0, 186.0, 304.0], '2cc84465-c146-490e-b427-588fd6c500e6': [188.0, 173.0, 359.0], '68c087da-91bb-46ca-bc22-15aee89605c9': [260.0, 246.0, 422.0], '44eb5a5d-3acc-40b5-84f9-39feae3c9926': [261.0, 245.0, 422.0], '64572e51-4a6d-4cbc-b02d-ee077a2948e3': [263.0, 251.0, 414.0], '47437c91-b2a0-4e6d-b65f-998a6af3383b': [276.0, 255.0, 397.0], 'edecf068-ecf5-4f14-85ff-969c82462f68': [221.0, 215.0, 376.0], 'e4aa6bc4-cc33-4778-ac4a-93da227ba08c': [218.0, 213.0, 381.0], '7a73b420-35ac-4493-a00a-a5735c5fbfb6': [560.0, 558.0, 1522.0]}
users after filtering= {'efec9f1b-fec5-4c08-bcd4-11522d67744f': [158.0, 139.0, 243.0], 'aaad230f-7349-4119-b8fb-aa3db9ec4dfa': [248.0, 226.0, 426.0], 'b07f1601-9e0f-4d80-8098-12d57e10b51c': [197.0, 150.0, 313.0], '38684b64-4680-4982-b22e-43d8343beaff': [1676.0, 1713.0, 4795.0], '0c21d656-b287-4e52-a50f-22818402b919': [187.0, 184.0, 349.0], '8f205586-b189-4929-bf52-c5f7d6754f8c': [164.0, 169.0, 327.0], 'ddfa7fa9-f5fa-4af8-a528-d9b731748920': [204.0, 159.0, 297.0], 'abc7e11e-fb5f-4674-9f26-9ac11eec9e09': [165.0, 144.0, 291.0], '23c0da7c-1e3f-4031-8b96-abc864b253c3': [178.0, 168.0, 314.0], '6bc81708-e173-435d-926f-d3030cbac13b': [170.0, 186.0, 304.0], '2cc84465-c146-490e-b427-588fd6c500e6': [188.0, 173.0, 359.0], '68c087da-91bb-46ca-bc22-15aee89605c9': [260.0, 246.0, 422.0], '44eb5a5d-3acc-40b5-84f9-39feae3c9926': [261.0, 245.0, 422.0], '64572e51-4a6d-4cbc-b02d-ee077a2948e3': [263.0, 251.0, 414.0], '47437c91-b2a0-4e6d-b65f-998a6af3383b': [276.0, 255.0, 397.0], 'edecf068-ecf5-4f14-85ff-969c82462f68': [221.0, 215.0, 376.0], 'e4aa6bc4-cc33-4778-ac4a-93da227ba08c': [218.0, 213.0, 381.0], '7a73b420-35ac-4493-a00a-a5735c5fbfb6': [560.0, 558.0, 1522.0]}
observed=
 [[ 158.  139.  243.]
 [ 248.  226.  426.]
 [ 197.  150.  313.]
 [1676. 1713. 4795.]
 [ 187.  184.  349.]
 [ 164.  169.  327.]
 [ 204.  159.  297.]
 [ 165.  144.  291.]
 [ 178.  168.  314.]
 [ 170.  186.  304.]
 [ 188.  173.  359.]
 [ 260.  246.  422.]
 [ 261.  245.  422.]
 [ 263.  251.  414.]
 [ 276.  255.  397.]
 [ 221.  215.  376.]
 [ 218.  213.  381.]
 [ 560.  558. 1522.]]
```

**Comment by vdbergh** [comment]
Thanks. I will try to check it.

**Comment by vdbergh** [comment]
I understand the issue. Due to numerical inaccuracies in scipy the residual is infinite. But an infinite residual is also an indicator of no residual.

I think the fix is to use sf/isf instead of cdf/ppf. This is in the latest commit.

**Comment by ppigazzini** [comment]
DEV updated, all residuals in the table.

**Comment by vdbergh** [comment]
I have squashed the PR and updated the commit message.

**Comment by vdbergh** [comment]
I have squashed the PR and updated the commit message.

File: server/fishtest/util.py [hunk 1]
```diff
@@ -105,62 +109,84 @@ def get_chi2(tasks, bad_users):
...
+def get_bad_workers(tasks, cached_chi2=None, p=0.001, res=7.0, iters=1):
+    # If we have an up-to-date result of get_chi2() we can pass
+    # it as cached_chi2 to avoid needless recomputation.
+    bad_workers = set()
+    for _ in range(iters):
+        if cached_chi2 is None:
+            chi2 = get_chi2(tasks, exclude_workers=bad_workers)
+        else:
+            chi2 = cached_chi2
+            chached_chi2 = None
```
Comment (ddobbelaere) [inline server/fishtest/util.py:147]: Typo, which might matter for iter>1

File: server/fishtest/util.py [hunk 2]
```diff
@@ -77,11 +79,13 @@ def get_chi2(tasks, bad_users):
         else:
             users[key] = wld
 
+    print("aggregated users=",users)
```
Comment (ppigazzini) [inline server/fishtest/util.py:82]: Debug leftover?

Comment (vdbergh) [reply server/fishtest/util.py]: Yep. I took it out in the latest push.

File: server/fishtest/util.py (hunk 1 above)
Comment (vdbergh) [reply server/fishtest/util.py]: Thanks. That was indeed a typo. I corrected it.

**Comment by vdbergh** [comment]
There is something strange with the coloring of the active workers:

https://dfts-0.pigazzini.it/tests/view/6153d50750986fceb8bd0145

I don't see this on the official site, so presumably it is not present in master.

**Comment by ppigazzini** [comment]
Yes, PROD is fine, I applied a dirty fix for the light theme, see #1052
The proper fix based on the new boostrap 5.1.1 table backfired on DEV for the dark theme :(

**Comment by ppigazzini** [comment]
@vdbergh too lazy to check the code. I'm wondering if the residual depends somehow also on the fraction of total games played by the worker, so that the residual value will change if we normalize the observed data to have each worker 1000 games played.

Since yesterday I'm testing the residuals changing the TC factor for one "bad worker"
https://github.com/glinscott/fishtest/blob/c7230ad5fc1b9b0763b213437bbbc575244b46a1/worker/games.py#L463-L466

The "bad worker" has a high residual when I divide factor by N. In this way the worker will play with a lower TC than expected and it will play more games than expected, see here factor /= 1.25 (base TC 1+0.01)
https://dfts-0.pigazzini.it/tests/view/615468aa21da0769d71caf34

The "bad worker" has a high residual only when I multiply factor with a big N. In this way the worker will play with a higher TC than expected and it will play less games than expected. With a lower N the residual is fine, see factor *= 1.25 (base TC 0.8+0.008)
https://dfts-0.pigazzini.it/tests/view/6154764521da0769d71caf3c

Here is the ongoing factor *= 2.0 (base TC 0.8+0.008)
https://dfts-0.pigazzini.it/tests/view/61548af921da0769d71caf42

**Comment by vdbergh** [comment]
Your result is a bit strange. Normally the worker that plays the most games should have the highest residual (in case the null hypothesis does not hold). Of course this assumes the same deviation of the null hypothesis. Perhaps this is the issue. At such fast time control *x is maybe not similar to /x.

**Comment by vdbergh** [comment]
I deleted my previous post since it was based upon an incorrect reading of your post.

So if I read correctly, if the bad worker plays more  games it gets a high residual  easily. If it plays fewer games then in your test it does not show a high residual. 

This is in some sense logical since there is less data available to base a conclusion upon. I suspect that with a longer test eventuality the bad worker will show a higher residual.

**Comment by ppigazzini** [comment]
OK, only to point out that time control is the same:
 - all workers @ 1+0.01 and bad worker @ 0.8+0.008
 - all workers @ 0.8+0.008 and bad worker @ 1+0.01

I'm wondering if makes sense to normalize the number of games by the workers to compute the residual.

**Comment by ppigazzini** [comment]
> This is in some sense logical since there is less data available to base a conclusion upon. I suspect that with a longer test eventuality the bad worker will show a higher residual.

Started 400k games with good workers @ 0.8+0.008 and bad worker @ 1.0+0.01:
https://dfts-0.pigazzini.it/tests/view/6154d5b821da0769d71caf54

Please mind that all good workers have 2 cores, the bad workers has 3 cores.

**Comment by vdbergh** [comment]
> OK, only to point out that time control is the same:
> 
> * good workers @ 1+0.01 and bad worker @ 0.8+0.008 - the bad workers get easily a high residual
> * good workers @ 0.8+0.008 and bad worker @ 1+0.01 - the bad worker get a normal residual
> 
> I'm wondering if makes sense to normalize the number of games played by the workers to compute the residual.

I think that multiplying a random variable with a number is a cosmetic operation. It does not essentially change the distribution.

What could possibly be done is to distribute the significant p-value (e.g. 0.05) differently based on the number of games of a worker. However I would need to think this through. I would prefer not to include this in this pr.

**Comment by ppigazzini** [comment]
It's a cosmetic operation if that don't change the residual distribution, in other terms if the residuals distribution depends upon the pentanomial probability distribution and not upon the total observed data.
linrock100cores should be purged if has a pentanomial probability distribution different from the other workers and not because it played way much games than the other workers.

BTW at 70k games the bad worker has residual 0.760 in the middle of the group. The bad worker played 6k games, one of the good worker 3k.

**Comment by vdbergh** [comment]
@ppigazzini It is unavoidable that the fewer games a worker has, the more difficult it is to prove (in a statistical sense) that it deviates from the null hypothesis (it is like measuring Elo). This is statistics. Multiplying the residuals with some computable number does not change this fact.

But I do not consider this a problem. It is exactly the workers with the most games that will influence a test the most. So it is the most important to detect these if they are bad.

In other words, if we have a bad worker with 200 games then this will be hard to detect (its residual will generally be green), but it does not matter a lot either.

I agree that the ungoing test is a bit funny. The chi2 test detects that there is an issue, but the residuals so far have been unable to pinpoint this on a particular worker.

The low p-value could simply mean that the 2 cores workers are not homogeneous without one being particularly worse than another. With the very fast time control this seems not unlikely to me.

**Comment by vdbergh** [comment]
As I said earlier. The correct statistical way to treat workers differently is obtained by distributing the p value non uniformly over the workers. But it needs some thought what exactly one wants to achieve. I am not convinced that there is a fundamental issue with uniform distribution (the Bonferroni correction).

**Comment by ppigazzini** [comment]
Please point out where I wrote to multiply the residual by a number.
I asked if the actual residual depends on the number of games played by the worker, the answer seems to be yes.
And I wrote if normalize the rows of the contingency table can help to pinpoint the bad worker, the implicit answer seems to be yes.
By the way the normalization of the row sum to 100 is a method used in statistics to spot up the differences (if any) in a contingency table.

**Comment by vdbergh** [comment]
@ppigazzini But whatever you do, you cannot change the fact the more games you have, the better conclusions you can draw. To bring up the analogy with Elo measurements once again. There is no statistical trick that will allow me to detect a difference of 1 Elo with say only a 1000 games. 

The same goes for detecting differences between pentanomial distributions. You simply need many games. So it is logical that the residual depends on the number of games (it depends of course also on the differences between the distributions).

But it seems to me that what you are really asking for is a summary table which shows the empirical pentanomial probabilities per worker (this is equivalent to normalising the row sums to be 100) so that you can visually inspect those probabilities. I am planning to implement exactly this. But the purpose of that would be more of heuristic nature, rather than being a hard statistical test.

**Comment by vdbergh** [comment]
> And I wrote if normalize the rows of the contingency table can help to pinpoint the bad worker, the implicit answer seems to be yes.

This I never said….

**Comment by vdbergh** [comment]
I thought a bit. A normalised row would no longer have a multinomial distribution. So you cannot for example use a contingency table with normalised rows as input to a chi2 test.

I suspect that the first step to use such a table would be to unnormalise the rows :)

**Comment by vdbergh** [comment]
Maybe the confusion is that you think that a worker automatically will get a high residual whenever it plays enough games. This is of course not true. But _if_ it deviates from the norm _and_ plays enough games, _then_ it will get a high residual. The number of games required depends on the magnitude of the deviation.

**Comment by ppigazzini** [comment]
Today I will be away from the keyboard until the evening, on mobile now :)

We want to know if H0 keeps for a test, so chi2 obviously must be computed with the observed data.

chi2 test can fail (H0 rejected) if a worker play a big chunk of the total games, the chi2 contribution is linear with the observed data if the cell.

In that case the Pearson residuals spot up the workers that contributed more to the chi2 (rejecting H0): non homogeneous empirical pentanomial frequencies and/or non homogenous games played.

But to auto purge a worker we should have an index to verify only if its empirical pentanomial probabilities are off from the mean one.


From the worker pentanomial observed data we can computed Elo (empirical pentanomial probabilities) and Elo interval (number of games), could this help?

**Comment by vdbergh** [comment]
> Today I will be away from the keyboard until the evening, on mobile now :)
> 
> We want to know if H0 keeps for a test, so chi2 obviously must be computed with the observed data.
> 
> chi2 test can fail (H0 rejected) if a worker play a big chunk of the total games, the chi2 contribution is linear with the observed data if the cell.
> 
> In that case the Pearson residuals spot up the workers that contributed more to the chi2 (rejecting H0): non homogeneous empirical pentanomial frequencies and/or non homogenous games played.
> 
> But to auto purge a worker we should have an index to verify only if its empirical pentanomial probabilities are off from the mean one.

But the problem is that you will only know the empirical pentanomial probabilities somewhat accurately for a worker with many games. For a worker with 100 games you may have empirical probabilities which strongly deviate from the norm, just for probabilistic reasons. This does not mean anything.

The current strategy is to purge the worker for which you have the most confidence that it deviates from the norm (the highest residual). One could perhaps instead try to purge the worker which is (provably) the farthest from the norm. It is not entirely clear to me how to implement this but maybe something reasonable is possible (but note any such strategy would only work with workers with "enough" games, so it is not clear what the benefit would be).

I any case I would prefer to leave pursuing alternative purging strategies out of this PR.

 

> 
> From the worker pentanomial observed data we can computed Elo (empirical pentanomial probabilities) and Elo interval (number of games), could this help?

**Comment by vdbergh** [comment]
Note that after initial hesitancy, the "bad worker" was eventually found in this test

https://dfts-0.pigazzini.it/tests/view/6154d5b821da0769d71caf54

**Comment by ppigazzini** [comment]
> Note that after initial hesitancy, the "bad worker" was eventually found in this test
> 
> https://dfts-0.pigazzini.it/tests/view/6154d5b821da0769d71caf54

Yes, only after that the colab workers stopped at 5-6 am and the bad worker played way more games than (some) of the colab workers. This morning at 6:30 am, with over 200k games played in the test, the bad worker was not red.

Purge requires a PhD in Categorical Data Analysis, auto purge is not an option :)

**Comment by ppigazzini** [comment]
> I any case I would prefer to leave pursuing alternative purging strategies out of this PR.

Of course, I'm simply sharing my views hoping this can somehow help to improve the chi2 test with one of these results: enable auto purge, guidelines to purge by hand, drop purge and use the chi2/residual only for analysis.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
PROD updated, thnak you @vdbergh :)

**Pull request merged**